File: 1248.numberOfSubArrays.cpp
class Solution {
   public:
    int numberOfSubarrays(vector<int>& nums, int k) {
        int l = 0;
        int r = 0;
        int cnt = 0;
        int n = nums.size();
        int ans = 0;

        while (r < n) {
            if (nums[r] % 2 == 1) {
                cnt++;
            }

            if (cnt == k) {
                int tmpr = r + 1;
                while (tmpr < n && nums[tmpr] % 2 != 1) {
                    tmpr++;
                }

                int tmpl = l;
                while (nums[tmpl] % 2 != 1) {
                    tmpl++;
                }

                ans += (tmpl - l + 1) * (tmpr - r);

                l = tmpl + 1;
                r = tmpr;
                cnt--;
                continue;
            }

            r++;
        }

        return ans;
    }
};

File: 1248.numberOfSubarrays.go
func numberOfSubarrays(nums []int, k int) int {
    n := len(nums)
	s := make([]int, n+1)
    cnt := make([]int, n+1)
    ans := 0
    
    cnt[s[0]]++;
    for i := 1; i <= n; i++ {
      s[i] = s[i-1] + nums[i-1]%2;
      cnt[s[i]]++;
    }
    
    for i := 1; i <= n; i++ {
      if s[i] >= k {
        ans += cnt[s[i]-k]
      }
    }

	return ans
}
